File: tools/packages/wesl-tooling/src/PackageNameUtils.ts
/** Package name sanitization for WESL.
 *
 * Converts typical npm package names to WGSL-safe identifiers using double-underscore encoding.
 * NPM package names can contain `@`, `/`, and `-`, which are not allowed in WGSL identifiers.
 *
 * ## Encoding Scheme
 *
 * ```
 * @     ==>  (remove)
 * /     ==>  __  (double underscore)
 * -     ==>  _   (single underscore)
 * ```
 *
 * ## Forward Mapping (npm ==> WGSL identifier)
 *
 * ```
 * my_package          ==>  my_package
 * random-wgsl         ==>  random_wgsl
 * @scope/my-pkg       ==>  scope__my_pkg
 * ```
 *
 * ## Reverse Mapping (WGSL identifier ==> npm package)
 *
 * ```
 * scope__my_pkg       ==>  try: @scope/my_pkg, @scope/my-pkg
 * random_wgsl         ==>  try: random_wgsl, random-wgsl
 * ```
 *
 * ## package.json Subpath Exports
 *
 * Subpaths don't create ambiguity because WeslBundle `name` field only contains package name:
 *
 * Scoped package:
 * ```
 * npm:        @foo/shader-utils
 * weslBundle: { name: "foo__shader_utils", modules: {...} }
 * WESL:       import foo__shader_utils::color::rgb2hsv
 * ```
 *
 * Unscoped package with subpath export:
 * ```
 * npm:        "foo"  (with exports: "./shader-utils": "./dist/...")
 * weslBundle: { name: "foo", modules: {...} }  // NOT foo__shader_utils!
 * WESL:       import foo::shader_utils::color::rgb2hsv  // Different identifier!
 * ```
 *
 * The `__` only appears when the package name itself contains `/`, never for subpaths.
 */

/** Convert npm package name to WGSL-safe identifier using double-underscore encoding. */
export function sanitizePackageName(npmName: string): string {
  return npmName
    .replace(/^@/, "") // Remove @ prefix
    .replaceAll("/", "__") // Replace / with __ (double underscore)
    .replaceAll("-", "_"); // Replace - with _ (single underscore)
}

/** Generate npm package name variations from sanitized WESL identifier.
 *
 * Uses double-underscore encoding to distinguish scoped vs unscoped packages:
 * - Has __ → scoped package (try @scope/pkg variants)
 * - No __ → unscoped package (try pkg variants)
 *
 * Examples:
 *   "lygia__shader_utils" → ["@lygia/shader_utils", "@lygia/shader-utils"]
 *   "random_wgsl" → ["random_wgsl", "random-wgsl"]
 */
export function* npmNameVariations(sanitizedPath: string): Generator<string> {
  const [pkg, sub] = breakAt(sanitizedPath, "/");

  let pkgName = pkg;
  let scopePrefix = "";

  if (pkg.includes("__")) {
    // presume a scoped npm package (@scope/pkg)
    const [scope, ...rest] = pkg.split("__");
    pkgName = rest.join("__"); // Rejoin in case of __ in package name (rare)
    scopePrefix = `@${scope}/`;
  }

  yield `${scopePrefix}${pkgName}${sub}`;
  yield `${scopePrefix}${pkgName.replaceAll("_", "-")}${sub}`;
}

/** Break string at first occurrence of delimiter.
 * @returns [before, after] where after includes the delimiter */
function breakAt(str: string, delimiter: string): [string, string] {
  const index = str.indexOf(delimiter);
  if (index === -1) return [str, ""];
  return [str.slice(0, index), str.slice(index)];
}

File: tools/npm-resolve-wesl/src/package_name_utils.rs
/// Generate npm package name variations from sanitized WESL identifier.
///
/// Uses double-underscore encoding to distinguish scoped vs unscoped packages:
/// - Has __ → scoped package (try @scope/pkg variants)
/// - No __ → unscoped package (try pkg variants)
///
/// Examples:
///   "lygia__shader_utils" → ["@lygia/shader_utils", "@lygia/shader-utils"]
///   "random_wgsl" → ["random_wgsl", "random-wgsl"]
pub fn npm_name_variations(sanitized_path: &str) -> Vec<String> {
    let (pkg, sub) = break_at(sanitized_path, "/");

    let (scope_prefix, pkg_name) = if pkg.contains("__") {
        // Scoped npm package (@scope/pkg)
        let parts: Vec<&str> = pkg.split("__").collect();
        let scope = parts[0];
        let name = parts[1..].join("__");
        (format!("@{}/", scope), name)
    } else {
        (String::new(), pkg.to_string())
    };

    vec![
        format!("{}{}{}", scope_prefix, pkg_name, sub),
        format!("{}{}{}", scope_prefix, pkg_name.replace('_', "-"), sub),
    ]
}

fn break_at(s: &str, delimiter: &str) -> (String, String) {
    if let Some(index) = s.find(delimiter) {
        (s[..index].to_string(), s[index..].to_string())
    } else {
        (s.to_string(), String::new())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_npm_name_variations() {
        let variations = npm_name_variations("random_wgsl");
        assert_eq!(variations, vec!["random_wgsl", "random-wgsl"]);

        let variations = npm_name_variations("lygia__shader_utils");
        assert_eq!(
            variations,
            vec!["@lygia/shader_utils", "@lygia/shader-utils"]
        );
    }

    #[test]
    fn test_break_at() {
        assert_eq!(
            break_at("foo/bar", "/"),
            ("foo".to_string(), "/bar".to_string())
        );
        assert_eq!(break_at("foo", "/"), ("foo".to_string(), String::new()));
    }
}
